app/http/httpclient.c

```diff
@@ -218,297 +219,297 @@ static int ICACHE_FLASH_ATTR http_chunked_decode( const char * chunked, char * d
 	}
 
-	if(req->headers == NULL) /* Avoid NULL pointer, it may cause exception */
-	{
+	if (req->headers == NULL) { /* Avoid NULL pointer, it may cause exception */
 		req->headers = (char *)os_malloc(sizeof(char));
 		req->headers[0] = '\0';
 	}
 
-	char buf[69 + strlen( req->method ) + strlen( req->path ) + strlen( req->hostname ) +
-		 strlen( req->headers ) + strlen( post_headers )];
-	int len = os_sprintf( buf,
-			      "%s %s HTTP/1.1\r\n"
-			      "Host: %s:%d\r\n"
-			      "Connection: close\r\n"
-			      "User-Agent: ESP8266\r\n"
-			      "%s"
-			      "%s"
-			      "\r\n",
-			      req->method, req->path, req->hostname, req->port, req->headers, post_headers );
-
-	if ( req->secure )
-		espconn_secure_send( conn, (uint8_t *) buf, len );
+	char buf[69 + strlen(req->method) + strlen(req->path) + strlen(req->hostname) +
+	         strlen(req->headers) + strlen(post_headers)];
+	int len = os_sprintf(buf,
+	                     "%s %s HTTP/1.1\r\n"
+	                     "Host: %s:%d\r\n"
+	                     "Connection: close\r\n"
+	                     "User-Agent: ESP8266\r\n"
+	                     "%s"
+	                     "%s"
+	                     "\r\n",
+	                     req->method, req->path, req->hostname, req->port, req->headers, post_headers);
+
+	if (req->secure)
+		espconn_secure_send(conn, (uint8_t *) buf, len);
 	else
-		espconn_send( conn, (uint8_t *) buf, len );
-	if(req->headers != NULL)
-		os_free( req->headers );
+		espconn_send(conn, (uint8_t *) buf, len);
+
+	if (req->headers != NULL)
+		os_free(req->headers);
+
 	req->headers = NULL;
-	HTTPCLIENT_DEBUG( "Sending request header\n" );
+	HTTPCLIENT_DEBUG("Sending request header\n");
 }
 
 
-static void ICACHE_FLASH_ATTR http_disconnect_callback( void * arg )
+static void ICACHE_FLASH_ATTR http_disconnect_callback(void * arg)
 {
-	HTTPCLIENT_DEBUG( "Disconnected\n" );
+	HTTPCLIENT_DEBUG("Disconnected\n");
 	struct espconn *conn = (struct espconn *) arg;
 
-	if ( conn == NULL )
-	{
+	if (conn == NULL) {
 		return;
 	}
 
-	if ( conn->proto.tcp != NULL )
-	{
-		os_free( conn->proto.tcp );
+	if (conn->proto.tcp != NULL) {
+		os_free(conn->proto.tcp);
 	}
-	if ( conn->reverse != NULL )
-	{
-		request_args_t	* req		= (request_args_t *) conn->reverse;
-		int		http_status	= -1;
-		char		* body		= "";
-
-		// Turn off timeout timer
-		os_timer_disarm( &(req->timeout_timer) );
-
-		if ( req->buffer == NULL )
-		{
-			HTTPCLIENT_DEBUG( "Buffer shouldn't be NULL\n" );
-		}
-		else if ( req->buffer[0] != '\0' )
-		{
+
+	if (conn->reverse != NULL) {
+		request_args_t  * req           = (request_args_t *) conn->reverse;
+		int             http_status     = -1;
+		char            * body          = "";
+
+		/* Turn off timeout timer */
+		os_timer_disarm(&(req->timeout_timer));
+
+		if (req->buffer == NULL) {
+			HTTPCLIENT_DEBUG("Buffer shouldn't be NULL\n");
+			/* To avoid NULL buffer */
+			req->buffer = "";
```

Comment:
We tested these on a special server, which always return 404 HTTP status code but have no body(no body means body length is 0). 
In this library the goal is "stable & easy to use", return NULL directly is OK but it will take extra work to user to check NULL pointer. 
In espressif's SDK, os_strlen("") is 0, but os_strlen(NULL) will cause a reboot due to read Illegal address. 
For example, if a response is 
`const char * resp;`
if it is NULL directly, user can't do such as os_printf("%s\n", resp), or will cause crash. User must check if(NULL != resp).
If a user just want's debug and print this response, they must use this style:
```
if(NULL == resp) {
    os_printf("");
} else {
    os_printf("%s", resp);
}
```
If using an empty string "" instead NULL, it just 
`    os_printf("%s", resp);`
So it can make user easy and reduced probability of bug when user forgot check this pointer.
That's why I changed this code, see also https://github.com/vowstar/esphttpclient/commit/5434b5178d9866e6203941c44acf6b00375c6009
This code fixed that when a server response NULL string, nodemcu firmware can handle it correctly and not crash to reboot.